[Global Opt] Add option to generalize matmul ops

IanWood1 opened the pull request:
In order to support -O* flags, generalizing matmul ops needs to be moved out of preprocessing and into global optimization. This adds a flag `iree-opt-generalize-matmul` and uses it during global optimization's `GeneralizeLinalgNamedOps` pass. Also, this changes SDXL tests to use this flag instead of the preprocessing pass.

MaheshRavishankar commented on compiler/src/iree/compiler/GlobalOptimization/GeneralizeLinalgNamedOps.cpp:
```diff
@@ -62,7 +63,12 @@ void GeneralizeLinalgNamedOpsPass::runOnOperation() {
   SmallVector<linalg::LinalgOp> namedOpCandidates;
   funcOp.walk([&](linalg::LinalgOp linalgOp) {
-    if (!IREE::Flow::isNonNullAndOutsideDispatch(linalgOp)) {
+    if (!IREE::Flow::isNonNullAndOutsideDispatch(linalgOp) ||
```
I dont think we need this here. This is in global optimizations, where there is no dispatches. So this seems unnecessary.

hanhanW replied:
We have such code in few GlobalOptimization passes. I thought that it is because people could pre-wrap dispatches? Is it still the case?

MaheshRavishankar replied:
Ah good point.

MaheshRavishankar commented on experimental/regression_suite/shark-test-suite-models/sdxl/test_unet.py:
```diff
@@ -218,7 +219,7 @@ def SDXL_PUNET_INT8_FP8_OUT(
 INT8_PUNET_FLAGS = [
     f"--iree-codegen-transform-dialect-library={iree_test_path_extension}/attention_and_matmul_spec_punet.mlir",
-    "--iree-preprocessing-pass-pipeline=builtin.module(util.func(iree-global-opt-raise-special-ops, iree-flow-canonicalize), iree-preprocessing-transpose-convolution-pipeline, iree-preprocessing-pad-to-intrinsics, util.func(iree-preprocessing-generalize-linalg-matmul-experimental))",
+    "--iree-preprocessing-pass-pipeline=builtin.module(util.func(iree-global-opt-raise-special-ops, iree-flow-canonicalize), iree-preprocessing-transpose-convolution-pipeline, iree-preprocessing-pad-to-intrinsics)",
```
Isnt the `iree-preprocessing-transpose-convolution-pipeline` the one that is converting from NCHW to NHWC convolutions?

IanWood1 committed:
Add option to generalize matmul ops

In order to support -O* flags, generalizing matmul ops needs to be moved out
of preprocessing and into global optimization. This adds a flag
`iree-opt-generalize-matmul` and uses it during global optimization's
`GeneralizeLinalgNamedOps` pass. Also, this changes SDXL tests to use this
flag instead of the preprocessing pass.

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>

hanhanW commented on compiler/src/iree/compiler/Pipelines/Options.h:
```diff
@@ -128,6 +128,9 @@ struct GlobalOptimizationOptions {
   // Strips debug assertions after any useful information has been extracted.
   bool stripAssertions = false;
 
+  // Strips debug assertions after any useful information has been extracted.
```
Please update the comment.

hanhanW commented on experimental/regression_suite/shark-test-suite-models/sdxl/test_unet.py:
```diff
@@ -218,7 +219,7 @@ def SDXL_PUNET_INT8_FP8_OUT(
 INT8_PUNET_FLAGS = [
     f"--iree-codegen-transform-dialect-library={iree_test_path_extension}/attention_and_matmul_spec_punet.mlir",
-    "--iree-preprocessing-pass-pipeline=builtin.module(util.func(iree-global-opt-raise-special-ops, iree-flow-canonicalize), iree-preprocessing-transpose-convolution-pipeline, iree-preprocessing-pad-to-intrinsics, util.func(iree-preprocessing-generalize-linalg-matmul-experimental))",
+    "--iree-preprocessing-pass-pipeline=builtin.module(util.func(iree-global-opt-raise-special-ops, iree-flow-canonicalize), iree-preprocessing-transpose-convolution-pipeline, iree-preprocessing-pad-to-intrinsics)",
```
I'm seeing it from experimental pass to global option. Is the long-term goal enabling it by default in IREE, so we add a global option for now?

IanWood1 replied:
Long term it would stay off by default but would be added to `-O2` ([further discussion on global opt flags](https://github.com/iree-org/iree/issues/19072)). I just need it out of preprocessing in order to enable it as a part of O2.


I think some back-ends struggle with generic ops.. but maybe if that would be fixed it could be default.

IanWood1 committed:
Fix comment

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>

IanWood1 committed:
Test removing RaiseSpecialOps from preprocessing

Signed-off-by: Ian Wood <ianwood2024@u.northwestern.edu>
